app/controllers/saml_completion_controller.rb

```diff
@@ -5,23 +5,35 @@
 class SamlCompletionController < ApplicationController
   # Pass the original service provider request to the main SamlIdpController#auth method
   # via a POST with form parameters replacing the url query parameters
+  before_action :verify_sp_session_exists
+
   def index
-    if sp_session.present?
-      request_url = URI(sp_session[:request_url])
-      path_year = request_url.path[-4..-1]
-      path_method = "api_saml_finalauthpost#{path_year}_url"
-      action_url = Rails.application.routes.url_helpers.send(path_method)
+    request_url = URI(sp_session[:request_url])
+    action_path = build_action_path(request_url.path)
 
+    if !action_path
+      render_not_found
+    else
       # Takes the query params which were set internally in the
       # sp_session (so they should always be valid).
       # A bad request that originated outside of the IDP would have
       # already responded with a 400 status before reaching this point.
       form_params = UriService.params(request_url)
 
-      render 'shared/saml_post_form', locals: { action_url: action_url, form_params: form_params },
+      render 'shared/saml_post_form', locals: { action_url: action_path, form_params: form_params },
                                       layout: false
-    else
-      render_not_found
     end
   end
+
+  def verify_sp_session_exists
+    render_not_found unless sp_session.present? && sp_session[:request_url].present?
+  end
+
+  def build_action_path(path)
+    path_year = path[-4..-1]
+    path = "/api/saml/finalauthpost#{path_year}"
+    recognized_path = Rails.application.routes.recognize_path(path, method: :post)
```

Comment:
TIL about this method 🌟

Reply:
Do we want to do any handling of a potential RoutingError here?

Reply:
in my testing, it didn't seem to raise since falls back to the `not_found` action:

```ruby
Rails.application.routes.recognize_path('fake_path', method: :post)
# => {:controller=>"pages", :action=>"page_not_found", :path=>"fake_path"}
```